src/lib.rs

```diff
@@ -39,4 +62,76 @@ mod tests {
         // THEN it should be zero
         assert_eq!(size, 0);
     }
-}
+
+    #[test]
+    fn it_should_report_the_correct_number_of_elements_when_filled() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 4 elements
+        for i in 0..4 {
+            buffer.add(i);    
+        }
+        let size = buffer.number_of_elements();
+        //then it should return 4
+        assert_eq!(size, 4);
+    }
+
+    #[test]
+    fn it_should_report_the_correct_number_of_elements_when_overfilled() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 12 elements
+        for i in 1..13 {
+            buffer.add(i);    
+        }
+        let size = buffer.number_of_elements();
+        //then it should return 10
+        assert_eq!(size, 10);
+    }
+
+    #[test]
+    fn it_should_return_the_before_pushed_values() {
+        //Given an empty buffer
+        let mut buffer: RingBuffer<i32> = RingBuffer::<i32>::new(10);
+        //when added 12 elements
+        for i in 1..11 {
+            println!("added {i}");
+            buffer.add(i);    
+        }
+        //then it should return 10
+        for i in 1..11 {
+            let ret_val = buffer.remove();
+            println!("expected {}, was {:?}", i , ret_val);
+            assert_eq!(ret_val, Some(i)); //Why has this to be some of...? 
```

Comment:
your ret_val's type is an Option<T>, so for rust this is it"s datatype.
This approach is fine, it is also correct to return an Option (since there are situations where you can't return anything). if you want to avoid the Some(i) here, you can unpack the ret_val when asserting on it:
```rust
assert_eq!(ret_val.unwrap(), i);
```

the `unwrap()` would panic (which means to quit your application) if ret_val was a None at this moment. This is ok in your testcode, you try to avoid `unwrap()` calls in your production code.